DOC: Improve documentation of ``as_array`` in ctypeslib.

TeusEE opened the pull request:
issue : [Issue No:#21335](https://github.com/numpy/numpy/issues/21335)

In ctypelib api doc, it will be helpful if object parameter have more rich explaination.

add some explain about obj parameter

please review it and apply it in API-Docs :)

TeusEE committed:
DOC: ctypeslib/as_array add desc about obj para

issue : https://github.com/numpy/numpy/issues/21335

In ctypelib api doc, it will be helpful if object parameter have more rich explaination.

add some explain about obj parameter

seberg commented:
Maybe say that it must be a pointer to a a single contiguous chunk of memory?  You could even link to the memory management section of the documentation in principle.
I just feel that to understand what this adds, you need to be in the right mind of state thinking about C static `double arr[4][4]` arrays?

Since we are touching this function, it would also be awesome to reformat it to `numpydoc` style; i.e. the same as some functions further up:
```
short description

Parmaters
---------
    obj : ...
        Pointer to a contiguous chunk of memory to be wrapped into an array.
    shape : int or tuple of ints
        Shape of the resulting array.
```
etc.

TeusEE commented:
> 

before the issue : https://github.com/numpy/numpy/issues/21335

I try to return 2d c array(built with c double pointer) from C object dll file.

because there are any explaination about obj parameter, i can't figure out how to convert returned data from C array to numpy ndarray
(in official docs, there are only mention about shape of obj Parameter)

i'll change style to numpydoc style(about shape and obj parameter) and PR again. thanks

TeusEE committed:
DOC: ctypeslib/as_array add desc with npdoc style

add API Docs about ctypeslib -> as_array function

change short desc -> full desc with numpydoc style

TeusEE committed:
DOC: ctypeslib/as_array add desc with npdoc style2

add API Docs about ctypeslib -> as_array function

change short desc -> full desc with numpydoc style

remove longlonglong sentence to long :(

TeusEE committed:
DOC: ctypeslib/as_array add desc with npdoc style3

add API Docs about ctypeslib -> as_array function

change short desc -> full desc with numpydoc style

remove longlonglong sentence to long :(2

TeusEE committed:
DOC: ctypeslib/as_array add desc with npdoc style3

add API Docs about ctypeslib -> as_array function

change short desc -> full desc with numpydoc style

remove longlonglong sentence to long :(

TeusEE committed:
DOC: ctypeslib/as_array add desc with npdoc style4

add API Docs about ctypeslib -> as_array function

change short desc -> full desc with numpydoc style

remove longlonglong sentence to long :(

+ change see Also link format

TeusEE committed:
DOC: ctypeslib/as_array add desc with npdoc style4

add API Docs about ctypeslib -> as_array function

change short desc -> full desc with numpydoc style

remove longlonglong sentence to long :(

+ change see Also link format

seberg commented on numpy/ctypeslib.py:
```diff
@@ -510,9 +510,34 @@ def as_array(obj, shape=None):
+        'ndarray' : numpy ndarray core structure
+            `<https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html>`_
+        Memory_Managment_Guide : 
+            `<https://numpy.org/devdocs/reference/c-api/data_memory.html>`_
```
Ah, that was hard to find, I was thinking this might be useful: https://numpy.org/devdocs/reference/arrays.ndarray.html#internal-memory-layout-of-an-ndarray but maybe this is also nice: https://numpy.org/devdocs/reference/c-api/array.html#c.PyArray_DATA

We mainly use "See Also" to refer to other functions that do a similar/related thing, so I think it would also be OK to juts include the links in the text under `Notes`.

Thanks for this though, it is nice to have this in numpydoc format.  Maybe, someone more involved in the Docs can have a look and help polish it up, or I will probably do in a bit.

TeusEE replied:
Adding link in docs with numpydoc style without checks error is hard to me
plz polish it up and close this pr :). @seberg .

TeusEE replied:
hmm... i don't know why this happen, but installing cygwin fail.
![image](https://user-images.githubusercontent.com/79349894/167255433-ec1eee74-8093-4af2-a8ce-aac97036c015.png)

TeusEE committed:
DOC: ctypeslib/as_array add desc with npdoc style5

add API Docs about ctypeslib -> as_array function

change short desc -> full desc with numpydoc style

remove longlonglong sentence to long :(

+ change see Also link format

+ change location http link from see also -> Notes

TeusEE committed:
DOC: ctypeslib/as_array add desc with npdoc style5

add API Docs about ctypeslib -> as_array function

change short desc -> full desc with numpydoc style

remove longlonglong sentence to long :(

+ change see Also link format

+ change location http link from see also -> Notes + change link page

charris committed:
MAINT: Grammar updates.

charris commented on numpy/ctypeslib.py:
```diff
@@ -509,10 +509,32 @@ def as_array(obj, shape=None):
+        Notes
+        -----                        
+        Numpy array stores its data as a void pointer in C. If you use a Pointer,
+        you must use ctypes Single Pointer.
```
What is a ctypes Single Pointer? There is no such documented ctype. And why is Pointer capitalized?

TeusEE replied:
@charris 
ctypes Single Pointer : i mean Python 'ctypes.POINTER'. 
[ctypes.pointer](https://docs.python.org/ko/3/library/ctypes.html#ctypes.pointer)
the reason why i use single is to emphasize not using double pointer
(because i using double pointer for numpy array data source before and have issue about it [issue](https://github.com/numpy/numpy/issues/21335))

Pointer capitalized : there are no mean using capitalized.
(i just think it can emphasize using pointer, but it might a bad choice. please amend to lower case ㅠㅠ)

seberg replied:
Sorry for not following up earlier.  I think the best solution w.r.t. the "Single Pointer" (which is indeed maybe not super clear without context), may be to rather create an
```
Examples
--------
```
section, and make an example for a 2-D array?

The example section is missing anyway and I suspect that it would be much clearer.

charris committed:
MAINT: Fix long line

seberg commented on numpy/ctypeslib.py:
```diff
@@ -509,10 +509,32 @@ def as_array(obj, shape=None):
+        PyArray_Data
+            `<https://numpy.org/devdocs/reference/c-api/array.html#c.PyArray_DATA>`_
+        Numpy_Array_Internal_Memory_layout
+            `<https://numpy.org/devdocs/reference/arrays.ndarray.html#internal-memory-layout-of-an-ndarray>`_
```
We should be able to reference these more directly for example:
```
NumPy stores its data as a single pointer in C (see :ref:`memory-layout` and :c:data:`PyArray_DATA`).
```
(I think, I did not test it.)

charris commented:
Thanks @TeusEE, I'm sorry this has been left hanging for so long. `numpy/ctypeslib.py` is now `numpy/ctypeslib/_ctypeslib.py`, so I'm going to close this. If you wish to continue, please open an new PR.
